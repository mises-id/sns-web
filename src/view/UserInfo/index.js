/*
 * @Author: lmk
 * @Date: 2021-07-15 12:51:04
 * @LastEditTime: 2021-08-26 13:34:51
 * @LastEditors: lmk
 * @Description: UserInfo page
 */
import Cell from '@/components/Cell';
import Image from '@/components/Image';
import { useBind } from '@/utils';
import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import {  Input ,Modal,Button,Picker,Toast} from 'zarm';
import './index.scss'
import Cropper from "react-cropper";
import "cropperjs/dist/cropper.css";
import { attachment } from '@/api/updata';
import { updateUser } from '@/api/user';
import { setUserInfo } from '@/utils/postMessage';
import { setLoginForm } from '@/actions/user';
import Navbar from '@/components/NavBar';
const genderList = [{ value: 'female', label: 'female' },
{ value: 'male', label: 'male' },{ value: 'other', label: 'other' }]
const UserInfo = (props)=>{
  const {t} = useTranslation()
  const {loginForm={}} = useSelector(state => state.user) || {}
  const [user, setuser] = useState(loginForm)
  const username = useBind(user.username)
  const phone = useBind(user.mobile)
  const mail = useBind(user.email)
  const address = useBind(user.address)
  const [avatar, setavatar] = useState('')
  const [visible, setvisible] = useState(false)
  const [cropper, setcropper] = useState(null)
  const [pickerVisible, setpickerVisible] = useState(false)
  const [selectGender, setselectGender] = useState([user.gender])
  const [avatarLoading, setavatarLoading] = useState(false)
  const [saveLoading, setsaveLoading] = useState(false)
  const dispatch = useDispatch()
  //set avatar
  const getAvatarChange = e=>{
    const file = e.target.files[0];
    if(file){
      const  reads= new FileReader();
      reads.readAsDataURL(file);
      reads.onload = ({target})=> {
        setavatar(target.result)
        e.target.value = ''
        setvisible(true)
      }
    }
  }
  //send avatar
  const sendAvatar = ()=>{
    const option = {
      width:200,
      height:200,
      fillColor: '#fff',
      imageSmoothingEnabled: false,
      imageSmoothingQuality: 'high',
    }
    cropper.getCroppedCanvas(option).toBlob((blob) => {
      const formData = new FormData();
      const filename = `avatar${user.uid}.${blob.type.replace('image/','')}`
      formData.append('file',blob,filename)
      formData.append('file_type','image')
      setavatarLoading(true)
      attachment(formData).then(res=>{
        setvisible(false);
        setuser(t=>{
          t.avatar = {};
          t.attachment_id = res.id;
          t.avatar.large = res.url;
          setAvatarInfo(res.id)
          return t;
        })
      }).finally(()=>setavatarLoading(false))
    })
  } 
  const pickerSubmit = e =>{
    if(e.length){
      user.gender = e[0].label;
      setuser({...user})
      setselectGender([e[0].value]);
    }
    setpickerVisible(false)
  }
  const saveInfo = (by="info")=>{
    setsaveLoading(true)
    const form = {
      username:{
        username:username.value
      },
      profile:{
        mobile:phone.value,
        email:mail.value,
        address:address.value,
        gender:user.gender
      }
    };
    if(by==='info'){
      const promise = [submit(form,'profile')];
      if(!loginForm.username){
        promise.push(submit(form,'username'))
      }
      Promise.all(promise).then(()=>{
        Toast.show(t('updataUserInfoSuccess'))
      })
    }
  }
  const setAvatarInfo = id=>{
    submit({avatar:{attachment_id:id}},'avatar').then(()=>{
      Toast.show(t('updataUserInfoSuccess'))
    })
  }
  const submit = async (form,by) =>{
    const byForm = {by};
    byForm[by] = form[by];
    try {
      const avatarDid = user.avatar&&user.avatar.large ? user.avatar.large : '';
      setUserInfo({username:username.value,gender:user.gender,mobile:phone.value,email:mail.value,avatarDid})
      const res = await updateUser(byForm)
      if(res){
        dispatch(setLoginForm(res))
        setsaveLoading(false)
        return Promise.resolve()
      }
      return errorMessage('error')
    } catch (error) {
      return errorMessage(error)
    }
  }
  const errorMessage = error=>{
    Toast.show(error)
    setsaveLoading(false)
    return Promise.reject(error)
  }
  return <div>
    <Navbar title={t('userInfoPageTitle')}/>
    <div className="m-layout m-bg-f8f8f8 userinfo">
      <div className="m-bg-fff m-padding-lr15">
        <Cell label={t('avatar')}  showIcon={false}
        rightChild={<div className="m-position-relative"><input type="file" accept="image/png,image/jpeg" className="avatar-input m-position-absolute" onChange={getAvatarChange}></input><Image size={35} source={user.avatar&&user.avatar.large}></Image></div>}></Cell>
        <Cell label={t('username')} showIcon={false}
        rightChild={<Input type="text" clearable={false} disabled={!!loginForm.username} {...username} placeholder={t('placeholder')} />}></Cell>
        <Cell label={t('gender')} showIcon={false}
        rightChild={<div className={!user.gender?'placeholder':''} onClick={()=>setpickerVisible(true)}>{user.gender || t('placeholder')}</div>}></Cell>
        <Cell label={t('phone')} showIcon={false}
        rightChild={<Input type="text" clearable={false} {...phone} placeholder={t('phonePlaceholder')} />}></Cell>
        <Cell label={t('mail')} showIcon={false}
        rightChild={<Input type="text" clearable={false} {...mail} placeholder={t('emailPlaceholder')} />}></Cell>
        <Cell label={t('address')} showIcon={false}
        rightChild={<Input type="text" clearable={false} {...address} placeholder={t('addressPlaceholder')} />}></Cell>
      </div>
      <div className="m-padding20 save-box">
        <Button block theme="primary" loading={saveLoading} ghost size="md" shape="round" onClick={()=>saveInfo('info')}>{t('send')}</Button>
      </div>
    </div>
    <Modal
      visible={visible}
      width="90%" 
      hasFooter={true}
      animationType="slideUp"
      footer={
        <>
          <div className="m-flex-1 m-padding-lr10">
            <Button block theme="primary" ghost size="sm" shape="round" onClick={()=>setvisible(false)}>
              {t('cancel')}
            </Button>
          </div>
          <div className="m-flex-1 m-padding-lr10">
            <Button block theme="primary" ghost loading={avatarLoading} size="sm" shape="round" onClick={sendAvatar}>
              {t('send')}
            </Button>
          </div>
        </>
      }
      maskClosable>
      <Cropper src={avatar}
        style={{ height: 400, width: "100%" }}
        onInitialized={setcropper}
       viewMode={1} minCropBoxWidth={100} minContainerHeight={300} minCropBoxHeight={100} initialAspectRatio={1} guides={false} aspectRatio={1}></Cropper>
    </Modal>
    <Picker visible={pickerVisible}
        dataSource={genderList}
        defaultValue={selectGender}
        onOk={pickerSubmit}
        onCancel={()=>setpickerVisible(false)}
        wheelDefaultValue={selectGender}
        value={selectGender}></Picker>
  </div>
}
export default UserInfo
