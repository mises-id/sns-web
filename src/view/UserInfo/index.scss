.userinfo{
  .za-input{
    padding-right:0;
    width: 50vw;
    font-size: 14px;
    color: #666;
    input{
      text-align: right;
    }
  }
  .avatar-input{
    width: 35px;
    height: 35px;
    z-index: 111;
    opacity: 0;
  }
}
.placeholder{
  color: #adadad;
}
.save-box{
  width: 80vw;
  margin: 50px auto 0;
}
body .za-button--primary.za-button--loading .za-activity-indicator.za-activity-indicator--circular circle{
  stroke:var(--theme-primary)
}
.NFT-container{
  padding:0 15px;
}
.NFT-select{
  &-items{
    grid-template-columns: 1fr 1fr 1fr;
    gap: 15px;
  }
  &-item{
    width: 100%;
    &-img{
      width: 100%;
      height: calc((100vw - 60px) / 3);
      border-radius: 10px;
      display: block;
    }
    &-img-box{
      margin-top: 18px;
      margin-bottom: 7px;
      position: relative;
      .active-image{
        transition: all 0.3s;
        opacity: 0;
        width: 25px;
        height: 25px;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        z-index: 33;
      }
      &.active{
        .active-image{
          opacity: 1;
        }
        &::after{
          content: "";
          width: calc(100% + 10px);
          height: calc(100% + 10px);
          border: 1px solid var(--button-primary-background);
          position: absolute;
          left: -5px;
          top: -5px;
          border-radius: 10px;
        }
        &::before{
          content: "";
          width:100%;
          height: 100%;
          background: rgba(0, 0, 0, 0.5);
          position: absolute;
          left: 0;
          top: 0;
          border-radius: 10px;
        }
      }
    }
    &-title{
      font-size: 12px;
      font-weight: 400;
      color: #333333;
    }
  }
  &-series-title{
    font-size: 19px;
    font-family: SF UI Display;
    font-weight: bold;
    margin-top: 25px;
    color: #333333;
  }
}
.pull-list{
  border: 1px solid var(--adm-border-color);
  .m-layout{
    height:calc(var(--window-height) - var(--window-layout-height) - 45px)
  }
}