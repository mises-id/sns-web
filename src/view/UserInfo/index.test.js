/*
 * @Author: lmk
 * @Date: 2021-07-15 12:51:10
 * @LastEditTime: 2021-07-15 12:51:20
 * @LastEditors: lmk
 * @Description: UserInfotest
 */
import React from 'react';
import UserInfo from './';
import {shallow } from 'enzyme';
describe('UserInfo', () => {
  it('should render correctly', () => {
    const component = shallow(<UserInfo />)
    expect(component).toMatchSnapshot();
  });
});