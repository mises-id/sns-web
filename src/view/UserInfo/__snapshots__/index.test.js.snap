// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo should render correctly 1`] = `
<div>
  <Navbar
    title="userInfoPageTitle"
  />
  <div
    className="m-layout m-bg-f8f8f8 userinfo"
  >
    <div
      className="m-bg-fff"
    >
      <Cell
        className="m-padding-lr15 m-padding-tb12"
        label="avatar"
        rightChild={
          <div
            className="m-position-relative"
            onClick={[Function]}
          >
            <input
              accept="image/png,image/jpeg"
              className="avatar-input hidden"
              onChange={[Function]}
              type="file"
            />
            <NFTAvatar
              size="35px"
            />
          </div>
        }
        showIcon={false}
      />
      <Cell
        className="m-padding-lr15 m-padding-tb19"
        label="username"
        rightChild={
          <Input
            clearable={false}
            maxLength={25}
            onChange={[Function]}
            placeholder="placeholder"
            type="text"
          />
        }
        showIcon={false}
      />
      <Cell
        className="m-padding-lr15 m-padding-tb19"
        label="gender"
        rightChild={
          <div
            className="placeholder"
            onClick={[Function]}
          >
            placeholder
          </div>
        }
        showIcon={false}
      />
      <Cell
        className="m-padding-lr15 m-padding-tb19"
        label="phone"
        rightChild={
          <Input
            className="userinfo-input"
            clearable={false}
            onChange={[Function]}
            placeholder="phonePlaceholder"
            type="text"
          />
        }
        showIcon={false}
      />
      <Cell
        className="m-padding-lr15 m-padding-tb19"
        label="mail"
        rightChild={
          <Input
            autoHeight={true}
            className="userinfo-input"
            clearable={false}
            onChange={[Function]}
            placeholder="emailPlaceholder"
            type="text"
          />
        }
        showIcon={false}
      />
      <Cell
        className="m-padding-lr15 m-padding-tb19 m-col-top"
        label="intro"
        rightChild={
          <div>
            <ForwardRef
              autoSize={
                Object {
                  "maxRows": 5,
                  "minRows": 3,
                }
              }
              defaultValue=""
              onChange={[Function]}
              placeholder="introPlaceholder"
              rows={2}
              showCount={false}
            />
          </div>
        }
        showIcon={false}
      />
    </div>
    <div
      className="m-padding20 save-box"
    >
      <ForwardRef
        block={true}
        color="primary"
        fill="outline"
        onClick={[Function]}
        shape="rounded"
      >
        Preview
      </ForwardRef>
      <ForwardRef
        block={true}
        className="m-margin-top10"
        color="primary"
        loading={false}
        onClick={[Function]}
        shape="rounded"
      >
        apply
      </ForwardRef>
    </div>
  </div>
  <Modal
    animationDuration={200}
    animationType="slideUp"
    closable={false}
    destroy={true}
    footer={
      <React.Fragment>
        <div
          className="m-flex-1 m-padding-lr10"
        >
          <ForwardRef
            block={true}
            color="primary"
            fill="outline"
            onClick={[Function]}
            shape="rounded"
            size="middle"
          >
            cancel
          </ForwardRef>
        </div>
        <div
          className="m-flex-1 m-padding-lr10"
        >
          <ForwardRef
            block={true}
            color="primary"
            loading={false}
            onClick={[Function]}
            shape="rounded"
            size="middle"
          >
            send
          </ForwardRef>
        </div>
      </React.Fragment>
    }
    hasFooter={true}
    mask={true}
    maskClosable={true}
    maskType="normal"
    prefixCls="za-modal"
    shape="radius"
    visible={false}
    width="83%"
  >
    <ForwardRef
      aspectRatio={1}
      guides={false}
      initialAspectRatio={1}
      minContainerHeight={300}
      minCropBoxHeight={100}
      minCropBoxWidth={100}
      onInitialized={[Function]}
      src=""
      style={
        Object {
          "height": 400,
          "width": "100%",
        }
      }
      viewMode={1}
    />
  </Modal>
  <ForwardRef(forwardRef)
    dataSource={
      Array [
        Object {
          "label": "female",
          "value": "female",
        },
        Object {
          "label": "male",
          "value": "male",
        },
        Object {
          "label": "other",
          "value": "other",
        },
      ]
    }
    defaultValue={
      Array [
        undefined,
      ]
    }
    onCancel={[Function]}
    onOk={[Function]}
    value={
      Array [
        undefined,
      ]
    }
    visible={false}
    wheelDefaultValue={
      Array [
        undefined,
      ]
    }
  />
  <ForwardRef(forwardRef)
    actions={
      Array [
        Object {
          "onClick": [Function],
          "text": "Avatar",
        },
        Object {
          "onClick": [Function],
          "text": "Avatar-NFT",
        },
      ]
    }
    onCancel={[Function]}
    onMaskClick={[Function]}
    spacing={true}
    visible={false}
  />
  <Popup
    bodyStyle={
      Object {
        "borderTopLeftRadius": "10px",
        "borderTopRightRadius": "10px",
      }
    }
    visible={false}
  >
    <NavBar
      backArrow={false}
      left={
        <span
          onClick={[Function]}
        >
          cancel
        </span>
      }
      right={
        <div
          style={
            Object {
              "display": "inline-block",
              "width": "61px",
            }
          }
        >
          <ForwardRef
            color="primary"
            onClick={[Function]}
            shape="rounded"
            size="small"
          >
            done
          </ForwardRef>
        </div>
      }
    />
    <div
      className="pull-list"
    >
      <ForwardRef(PullList)
        data={Array []}
        emptyText="NFTEmpty"
        load={[Function]}
        renderView={[Function]}
      />
    </div>
  </Popup>
</div>
`;
